`hardhat compile` randomly and quietly stops before compiling contracts

Edit by @fvictorio: this is caused by a bug in Node v18.16.0. For a temporary workaround, downgrade your Node version to 18.15.0.

---

### Version of Hardhat

2.14.0

### What happened?

`npx hardhat compile` sometimes goes on with compiling contracts and sometimes (seems totally random) silently stops at the `Downloading compiler ...` step before any contracts are actually compiled. This issue is reproducible with no contracts already compiled and in docker.

### Minimal reproduction steps

Here's a simple repository where you can observe the problem: [repo](https://github.com/sebastiendan/hardhat-compile-issue).

This CI workflow run shows in the logs that the compilation didn't happen: [log](https://github.com/sebastiendan/hardhat-compile-issue/actions/runs/4788710757/jobs/8515780386#step:7:271).

In addition, one can run the docker container in interactive mode and manually run the `build` command to see that, again, the command sometimes doesn't build nor throws an error but simply quietly stops without compiling (same output as the CI log above):
- Run the container in interactive mode: `docker run --rm -it ghcr.io/sebastiendan/hardhat-compile-issue:main bash`
- Run the compile command: `npm run build`
- Observe sometimes an actual compilation, sometimes a quiet stop in the command execution
(run the above multiple times by exiting the container and re-running the two commands above)

### Search terms

no compilation

I'm getting this error (more or less, I think?), also reported in https://github.com/NomicFoundation/hardhat/issues/2569

Here's the Github Actions run: https://github.com/streamr-dev/network-contracts/actions/runs/5088395393/jobs/9144745279?pr=523#step:6:15

Explanation:
* `rm -rf artifacts && hardhat compile && rm -rf dist && tsc -p tsconfig.build.json`
  * I try to force compilation by deleting `artifacts` first (to make double sure there isn't an existing artifacts directory, so hardhat might skip compilation)
  * after, I delete `dist` just in case, and run `tsc` that outputs to `dist`
* `src/exports.ts(1,66): error TS2307: Cannot find module '../artifacts/contracts/OperatorTokenomics/Operator.sol/Operator.json' or its corresponding type declarations`
  * `tsc` says: nothing found in `artifacts`. Right after `hardhat compile`.

What makes this problem frustrating is that
* it doesn't happen every time in Github Actions
* it doesn't happen much if at all locally

--- a/.changeset/tricky-parents-warn.md
+++ b/.changeset/tricky-parents-warn.md
@@ -1,0 +1,5 @@
+---
+"hardhat": patch
+---
+
+Fixed an issue related to compiler downloads in node v18.16.x

--- a/packages/hardhat-core/src/internal/util/download.ts
+++ b/packages/hardhat-core/src/internal/util/download.ts
@@ -1,9 +1,7 @@
 import type { Dispatcher } from "undici";
 
-import fs from "fs";
 import fsExtra from "fs-extra";
 import path from "path";
-import util from "util";
 
 import { getHardhatVersion } from "./packageInfo";
 import { shouldUseProxy } from "./proxy";
@@ -26,9 +24,7 @@ export async function download(
   timeoutMillis = 10000,
   extraHeaders: { [name: string]: string } = {}
 ) {
-  const { pipeline } = await import("stream");
   const { getGlobalDispatcher, ProxyAgent, request } = await import("undici");
-  const streamPipeline = util.promisify(pipeline);
 
   let dispatcher: Dispatcher;
   if (process.env.http_proxy !== undefined && shouldUseProxy(url)) {
@@ -52,10 +48,11 @@ export async function download(
   });
 
   if (response.statusCode >= 200 && response.statusCode <= 299) {
+    const responseBody = await response.body.arrayBuffer();
     const tmpFilePath = resolveTempFileName(filePath);
     await fsExtra.ensureDir(path.dirname(filePath));
 
-    await streamPipeline(response.body, fs.createWriteStream(tmpFilePath));
+    await fsExtra.writeFile(tmpFilePath, responseBody);
     return fsExtra.move(tmpFilePath, filePath, { overwrite: true });
   }
   // undici's response bodies must always be consumed to prevent leaks

